packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java

```diff
@@ -0,0 +1,363 @@
+import android.app.KeyguardManager;
+import android.os.StatFs;
+import android.telephony.TelephonyManager;
+import android.app.ActivityManager;
+import android.app.UiModeManager;
+import android.view.WindowManager;
+import android.util.DisplayMetrics;
+import android.content.res.Configuration;
+import android.Manifest;
+
+import org.unimodules.core.ExportedModule;
+import org.unimodules.core.ModuleRegistry;
+import org.unimodules.core.Promise;
+import org.unimodules.core.interfaces.ActivityProvider;
+import org.unimodules.core.interfaces.ExpoMethod;
+import org.unimodules.core.interfaces.RegistryLifecycleListener;
+
+import java.math.BigInteger;
+import java.net.InetAddress;
+import java.net.NetworkInterface;
+import java.nio.ByteOrder;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class DeviceModule extends ExportedModule implements RegistryLifecycleListener, ActivityCompat.OnRequestPermissionsResultCallback {
+  private static final String NAME = "ExpoDevice";
+  private static final String TAG = DeviceModule.class.getSimpleName();
+  private final int REQUEST_READ_PHONE_STATE = 1;
+
+  private ModuleRegistry mModuleRegistry;
+  private Context mContext;
+  private WifiInfo wifiInfo;
+  private ActivityProvider mActivityProvider;
+  private Activity mActivity;
+  private DeviceType mDeviceType;
+  private String mPhoneNumber;
+
+  public DeviceModule(Context context) {
+    super(context);
+    mContext = context;
+  }
+
+  public enum DeviceType {
+    HANDSET("Handset"),
+    TABLET("Tablet"),
+    TV("Tv"),
+    UNKNOWN("Unknown");
+
+    private final String value;
+
+    DeviceType(String value) {
+      this.value = value;
+    }
+
+    public String getValue() {
+      return value;
+    }
+  }
+
+
+  private WifiInfo getWifiInfo() {
+    WifiManager manager = (WifiManager) mContext.getApplicationContext().getSystemService(Context.WIFI_SERVICE);
+    this.wifiInfo = manager.getConnectionInfo();
+    return this.wifiInfo;
+  }
+
+  @Override
+  public String getName() {
+    return NAME;
+  }
+
+  @Override
+  public void onCreate(ModuleRegistry moduleRegistry) {
+    mModuleRegistry = moduleRegistry;
+    mActivityProvider = moduleRegistry.getModule(ActivityProvider.class);
+    mActivity = mActivityProvider.getCurrentActivity();
+  }
+
+  @Override
+  public void onRequestPermissionsResult(int requestCode,
+                                         String[] permissions, int[] grantResults) {
+    switch (requestCode) {
+      case REQUEST_READ_PHONE_STATE: {
+        // If request is cancelled, the result arrays are empty.
+        if (grantResults.length > 0
+            && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+            try{
+              TelephonyManager telMgr = (TelephonyManager) mContext.getApplicationContext().getSystemService(Context.TELEPHONY_SERVICE);
+              mPhoneNumber = telMgr.getLine1Number();
```

Comment:
In order to communicate with the promise in `getPhoneNumberAsync` in `onRequestPermissionsResult, I would throw an `AccessControlException` so that in `getPhoneNumberAsync` I would catch it and reject the promise if failed to acquire the permission.

I tried to use `permissionsModule.askForPermissions` in `ImagePickerModule`, but got into `InvocationTargetException` every time I tried to acquire permission. Couldn't figure that out, so I used throw exception to reject the promise.

Reply:
discussed offline, commenting here for reference 🙂  - need to add this permission to PermissionsModule